dali/driver/hasseb.py

```diff
@@ -34,160 +50,410 @@ def __repr__(self):
-        else:
-            self._pending = None
-        self.backend.write(data)
+    #def __init__(self, processEvents):
+    #    self._processEvents = processEvents
 
-    def receive(self, data):
-        frame = self.extract(data)
-        if isinstance(frame, HassebDALIUSBNoDataAvailable):
-            return
-        elif isinstance(frame, BackwardFrame):
-            if self._pending:
-                command, callback, kw = self._pending
-                callback(command.response(frame), **kw)
+    def setEventHandler(self, processEvents):
+        self._processEvents = processEvents
+
+    def wait_for_response(self):
+        """Wait for response message. Timeout 2000 ms.
+        """
+        for i in range(200):
+            if not self._pending:
+                return
             else:
-                logger.error("Received frame for no pending command")
-        else:
-            logger.error("Received frame is not BackwardFrame")
-
-
-def _test_async(logger, command):
-    print('Test async driver')
-    driver = AsyncHassebDALIUSBDriver()
-    driver.logger = logger
-
-    # async response callback
-    def response_received(response):
-        print('Response received: {}'.format(response))
-
-    driver.send(command, callback=response_received)
-
-    # exit callback
-    def signal_handler(signal, frame):
-        driver.backend.close()
-        sys.exit(0)
-
-    import signal
-    signal.signal(signal.SIGINT, signal_handler)
-    print('Press Ctrl+C')
-    signal.pause()
-
-def _test_sync(logger, command):
-    print('Test sync driver')
-    driver = SyncHassebDALIUSBDriver()
-    driver.logger = logger
-
-    print('Response: {}'.format(driver.send(command)))
-    driver.backend.close()
-
-if __name__ == '__main__':
-    """Usage: python tridonic.py sync|async address value."""
-    from dali.gear.general import DAPC, QueryActualLevel
-    import sys
-
-    # setup console logging
-    logger = logging.getLogger('HassebDALIDriver')
-    logger.setLevel(logging.DEBUG)
-    handler = logging.StreamHandler(sys.stdout)
-    handler.setLevel(logging.DEBUG)
-    handler.setFormatter(logging.Formatter(
-        '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
-    ))
-    logger.addHandler(handler)
-
-    # command to send
-    cmd = DAPC(int(sys.argv[2]), int(sys.argv[3]))
-
-    # sync interface
-    if sys.argv[1] == 'sync':
-        _test_sync(logger, cmd)
-    # async interface
-    elif sys.argv[1] == 'async':
-        _test_async(logger, cmd)
+                self._processEvents()
+                time.sleep(0.01)
+
+
+class SyncHassebDALIUSBDriver(HassebDALIUSBDriver, SyncDALIDriver):
+    """Synchronous ``DALIDriver`` implementation for Hasseb DALI USB device.
+    """
+
+    def wait_for_response(self):
+        """Wait for response message.
+        """
+        for i in range(200):
+            if not self._pending:
+                return
+            else:
+                self.receive()
+        
```

Comment:
All the code below this line is device-independent and doesn't belong in a driver module.

Reply:
Yes, but there are two issues:
1. In the bus.py you do not recommend to use it
2. QT, which I'm using in the graphical user interface, don't like exceptions, so I had to remove exceptions.

Reply:
1. You could put it in the GUI project.
2. If the driver raises exceptions, they should be caught by the GUI code (and suitable action taken) before they propagate up the call stack to any Qt code.

Reply:
Thanks I will solve this somehow, it is required only for the hasseb_test.py. I will use your bus.py in the test and make my own implementation to the gui folder.

Reply:
Or do you prefer removing the hasseb_test.py?

Reply:
My plan for removing `bus.py` and replacement code that's driver independent is in pull request #68.